### ISSUE TITLE
Fonts: User-specified letter spacing

### ISSUE THREAD

**Issue by rozbb** [wez/wezterm#3554]
### Is your feature request related to a problem? Please describe.

Some fonts look much better with a tighter letter spacing (aka "horizontal spacing" aka "tracking"). Here is the same screen in iTerm2, with 100% letterspacing and then 88% letterspacing using Fira Code SemiBold

<img width="1103" alt="Screenshot 2023-04-18 at 00 38 53" src="https://user-images.githubusercontent.com/752802/232673670-66656825-07c4-40a1-84ac-45518360ffd4.png">
<img width="994" alt="Screenshot 2023-04-18 at 00 38 04" src="https://user-images.githubusercontent.com/752802/232673682-c2f56fe2-0763-45df-8020-0a54e8769964.png">

### Describe the solution you'd like

iTerm2 lets you set the letterspacing as a percentage (see below). For wezterm, I think it'd make sense to have a float-valued `horizontal_spacing` key in `wezterm.font`'s `attribute` parameter. E.g.,

```lua
config.font = wezterm.font('Fira Code', { weight = 'SemiBold', horizontal_spacing = 0.88 })
```

<img width="1055" alt="Screenshot 2023-04-18 at 00 53 11" src="https://user-images.githubusercontent.com/752802/232674663-48c2984f-43d5-4372-94a5-a3d63367289d.png">

### Describe alternatives you've considered

I've looked through all the font size, weight, stretch, and shaping options, and none cover this particular case. Not every font has a Condensed version, for example.

**Comment by wez** [comment]
Did you find https://wezfurlong.org/wezterm/config/lua/config/cell_width.html ?

There are limitations with this sort of thing because wezterm doesn't know directly what constitutes "a letter".  The harfbuzz library is responsible for shaping sequences of codepoints into sequences of glyphs based on mapping tables defined by the font. Ligatured glyph sequences can be produced from 1 or more codepoints in a way where it is difficult to reason about what humans would perceive to be letter boundaries. There is not a "letter" concept in the font shaping api.

For example, it is common for the sequence `->` to be replaced by a ligature for a right arrow. There are at least three variations on this that I have seen:

* Alternative `-` and `>` glyphs are substituted by the font that have almost the same monospace widths as the non-ligatured versions
* An empty `-` and a double wide special `->` glyph are substituted. Or a double wide `->` followed by an empty `>` glyph.
* A single double width `->` is substituted

It is very difficult for wezterm to reason about what is a letter here because there are two input codepoints and the output appears (when rendered) to be a single glyph.  Only one of these cases has an "atomic" single glyph output from the shaper that you might consider to be subject to revised spacing.  The first two cases have two output glyphs, and it would be invalid to change their spacing because it would mess up the appearance of the ligature for the second case.

This stuff is further complicated when BIDI comes into play, which can change the order of the rendered glyphs for right-to-left script sub-sequences.

The `cell_width` option implements a very basic and limited form of this kind of spacing control: it is blissfully unaware of ligatures.

I'm not eager to see this concept applied to individual fonts because that would require a much more involved implementation and revising how cell width is computed overall when determining the terminal matrix dimensions.
This is something that I can see being a bug magnet with a high ongoing maintenance cost to some already complex code.

So, in short, I think `cell_width` is likely as good as you're going to get for now.

**Comment by rozbb** [comment]
Doh, I don't know how I missed that! Yes, I think this solves the bulk of my issue. Sorry for the bother!
